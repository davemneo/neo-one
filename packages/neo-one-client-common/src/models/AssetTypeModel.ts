import { InvalidAssetTypeError, InvalidAssetTypeJSONError } from '../errors';
import { AssetTypeJSON } from './types';

export enum AssetTypeModel {
  CreditFlag = 0x40,
  DutyFlag = 0x80,
  GoverningToken = 0x00,
  UtilityToken = 0x01,
  Currency = 0x08,
  Share = 0x90,
  Invoice = 0x98,
  Token = 0x60,
}

export const hasFlag = (
  assetType: AssetTypeModel,
  flag: AssetTypeModel,
  // tslint:disable-next-line
): boolean => (assetType & flag) === flag;

const isAssetType = (assetType: number): assetType is AssetTypeModel =>
  // tslint:disable-next-line strict-type-predicates
  AssetTypeModel[assetType] !== undefined;

export const assertAssetType = (assetType: number): AssetTypeModel => {
  if (!isAssetType(assetType)) {
    throw new InvalidAssetTypeError(assetType);
  }

  return assetType;
};

export const toJSONAssetType = (type: AssetTypeModel): AssetTypeJSON => assertAssetTypeJSON(AssetTypeModel[type]);

const isAssetTypeJSON = (assetType: string): assetType is AssetTypeJSON =>
  // tslint:disable-next-line strict-type-predicates no-any
  AssetTypeModel[assetType as any] !== undefined;

export const assertAssetTypeJSON = (type: string): AssetTypeJSON => {
  if (!isAssetTypeJSON(type)) {
    throw new InvalidAssetTypeJSONError(type);
  }

  return type;
};

export const toAssetType = (type: AssetTypeJSON): AssetTypeModel => AssetTypeModel[type];
