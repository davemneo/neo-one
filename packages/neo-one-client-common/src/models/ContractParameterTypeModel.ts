import { InvalidContractParameterTypeError, InvalidContractParameterTypeJSONError } from '../errors';
import { ContractParameterTypeJSON } from './types';

export enum ContractParameterTypeModel {
  Signature = 0x00,
  Boolean = 0x01,
  Integer = 0x02,
  Hash160 = 0x03,
  Hash256 = 0x04,
  ByteArray = 0x05,
  PublicKey = 0x06,
  String = 0x07,
  Array = 0x10,
  Map = 0x12,
  InteropInterface = 0xf0,
  Void = 0xff,
}

const isContractParameterType = (value: number): value is ContractParameterTypeModel =>
  // tslint:disable-next-line strict-type-predicates
  ContractParameterTypeModel[value] !== undefined;

export const assertContractParameterType = (value: number): ContractParameterTypeModel => {
  if (isContractParameterType(value)) {
    return value;
  }
  throw new InvalidContractParameterTypeError(value);
};

export const toJSONContractParameterType = (type: ContractParameterTypeModel): ContractParameterTypeJSON =>
  assertContractParameterTypeJSON(ContractParameterTypeModel[type]);

const isContractParameterTypeJSON = (value: string): value is ContractParameterTypeJSON =>
  // tslint:disable-next-line strict-type-predicates no-any
  ContractParameterTypeModel[value as any] !== undefined;

export const assertContractParameterTypeJSON = (value: string): ContractParameterTypeJSON => {
  if (isContractParameterTypeJSON(value)) {
    return value;
  }
  throw new InvalidContractParameterTypeJSONError(value);
};

export const toContractParameterType = (value: ContractParameterTypeJSON): ContractParameterTypeModel =>
  assertContractParameterType(ContractParameterTypeModel[value]);
