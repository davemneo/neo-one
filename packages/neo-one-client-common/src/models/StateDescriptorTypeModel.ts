import { InvalidStateDescriptorTypeError } from '../errors';

export enum StateDescriptorTypeModel {
  Account = 0x40,
  Validator = 0x48,
}

const isStateDescriptorType = (value: number): value is StateDescriptorTypeModel =>
  // tslint:disable-next-line strict-type-predicates
  StateDescriptorTypeModel[value] !== undefined;

export const assertStateDescriptorType = (value: number): StateDescriptorTypeModel => {
  if (isStateDescriptorType(value)) {
    return value;
  }
  throw new InvalidStateDescriptorTypeError(value);
};
