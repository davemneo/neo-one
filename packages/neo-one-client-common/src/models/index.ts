export * from './AssetTypeModel';
export * from './ContractParameterTypeModel';
export * from './Serializable';
export * from './StateDescriptorTypeModel';
export * from './StorageFlagsModel';
export * from './WitnessModel';
export * from './transaction';
export * from './types';
export * from './vm';
