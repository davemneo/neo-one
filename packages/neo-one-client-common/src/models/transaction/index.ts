export * from './attribute';
export * from './ClaimTransactionModel';
export * from './InputModel';
export * from './InvocationTransactionModel';
export * from './OutputModel';
export * from './TransactionBaseModel';
export * from './TransactionModel';
export * from './TransactionTypeModel';
