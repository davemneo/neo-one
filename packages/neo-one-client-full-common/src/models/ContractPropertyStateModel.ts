import { InvalidContractPropertyStateError } from '../errors';

export enum ContractPropertyStateModel {
  NoProperty = 0x00,
  HasStorage = 0x01,
  HasDynamicInvoke = 0x02,
  Payable = 0x04,
  HasStorageDynamicInvoke = 0x03,
  HasStoragePayable = 0x05,
  HasDynamicInvokePayable = 0x06,
  HasStorageDynamicInvokePayable = 0x07,
}

// tslint:disable-next-line variable-name
export const HasStorage = new Set([
  ContractPropertyStateModel.HasStorage,
  ContractPropertyStateModel.HasStorageDynamicInvoke,
  ContractPropertyStateModel.HasStoragePayable,
  ContractPropertyStateModel.HasStorageDynamicInvokePayable,
]);

// tslint:disable-next-line variable-name
export const HasDynamicInvoke = new Set([
  ContractPropertyStateModel.HasDynamicInvoke,
  ContractPropertyStateModel.HasStorageDynamicInvoke,
  ContractPropertyStateModel.HasDynamicInvokePayable,
  ContractPropertyStateModel.HasStorageDynamicInvokePayable,
]);

// tslint:disable-next-line variable-name
export const HasPayable = new Set([
  ContractPropertyStateModel.Payable,
  ContractPropertyStateModel.HasStoragePayable,
  ContractPropertyStateModel.HasDynamicInvokePayable,
  ContractPropertyStateModel.HasStorageDynamicInvokePayable,
]);

const isContractPropertyState = (value: number): value is ContractPropertyStateModel =>
  // tslint:disable-next-line strict-type-predicates
  ContractPropertyStateModel[value] !== undefined;

export const assertContractPropertyState = (value: number): ContractPropertyStateModel => {
  if (isContractPropertyState(value)) {
    return value;
  }

  throw new InvalidContractPropertyStateError(value);
};

export const getContractProperties = ({
  hasStorage,
  hasDynamicInvoke,
  payable,
}: {
  readonly hasStorage: boolean;
  readonly hasDynamicInvoke: boolean;
  readonly payable: boolean;
}): ContractPropertyStateModel => {
  if (hasStorage && hasDynamicInvoke && payable) {
    return ContractPropertyStateModel.HasStorageDynamicInvokePayable;
  }

  if (hasStorage && hasDynamicInvoke) {
    return ContractPropertyStateModel.HasStorageDynamicInvoke;
  }

  if (hasStorage && payable) {
    return ContractPropertyStateModel.HasStoragePayable;
  }

  if (hasDynamicInvoke && payable) {
    return ContractPropertyStateModel.HasDynamicInvokePayable;
  }

  if (hasDynamicInvoke) {
    return ContractPropertyStateModel.HasDynamicInvoke;
  }

  if (hasStorage) {
    return ContractPropertyStateModel.HasStorage;
  }

  if (payable) {
    return ContractPropertyStateModel.Payable;
  }

  return ContractPropertyStateModel.NoProperty;
};
