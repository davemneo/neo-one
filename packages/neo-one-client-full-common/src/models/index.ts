export * from './BaseState';
export * from './ContractModel';
export * from './ContractPropertyStateModel';
export * from './transaction';
