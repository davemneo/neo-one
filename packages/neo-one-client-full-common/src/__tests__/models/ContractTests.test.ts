import { BinaryWriter } from '@neo-one/client-common';
import { assertContractPropertyState, ContractModel, getContractProperties } from '../../models';

describe('Contract Model Tests', () => {
  test('Serialize Wire Base', () => {
    const writer = new BinaryWriter();
    const contractModel = new ContractModel({
      script: Buffer.alloc(25),
      parameterList: [],
      returnType: 0x01,
      contractProperties: 0x01,
      name: 'test',
      codeVersion: '0.0.0',
      author: 'Ben Dover',
      email: 'example@neo-one.gov',
      description: 'test',
    });

    contractModel.serializeWireBase(writer);
    expect(writer.buffer).toMatchSnapshot();
  });
});

describe('Contract Model Property State Tests', () => {
  const badStateValue = 0x10;

  const createBooleanContractProperties = (storageIn?: boolean, invokeIn?: boolean, payableIn?: boolean) => ({
    hasStorage: storageIn === undefined ? false : storageIn,
    hasDynamicInvoke: invokeIn === undefined ? false : invokeIn,
    payable: payableIn === undefined ? false : payableIn,
  });

  test('assertPropertyState throws on bad value', () => {
    expect(() => assertContractPropertyState(badStateValue)).toThrowError(
      `Expected contract parameter type, found: ${badStateValue.toString(16)}`,
    );
  });

  test('getContractProperties - all', () => {
    expect(getContractProperties(createBooleanContractProperties(true, true, true))).toEqual(0x07);
  });
  test('getContractProperties - storage, invoke', () => {
    expect(getContractProperties(createBooleanContractProperties(true, true))).toEqual(0x03);
  });
  test('getContractProperties - invoke, payable', () => {
    expect(getContractProperties(createBooleanContractProperties(false, true, true))).toEqual(0x06);
  });
  test('getContractProperties - invoke', () => {
    expect(getContractProperties(createBooleanContractProperties(false, true))).toEqual(0x02);
  });
  test('getContractProperties - storage', () => {
    expect(getContractProperties(createBooleanContractProperties(true))).toEqual(0x01);
  });
  test('getContractProperties - payable', () => {
    expect(getContractProperties(createBooleanContractProperties(false, false, true))).toEqual(0x04);
  });
  test('getContractProperties - no property', () => {
    expect(getContractProperties(createBooleanContractProperties())).toEqual(0x00);
  });
});
