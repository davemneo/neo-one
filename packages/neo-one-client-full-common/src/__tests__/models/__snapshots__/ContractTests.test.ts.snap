// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contract Model Tests Serialize Wire Base 1`] = `
Array [
  "19",
  "00000000000000000000000000000000000000000000000000",
  "00",
  "",
  "01",
  "01",
  "04",
  "74657374",
  "05",
  "302e302e30",
  "09",
  "42656e20446f766572",
  "13",
  "6578616d706c65406e656f2d6f6e652e676f76",
  "04",
  "74657374",
]
`;
